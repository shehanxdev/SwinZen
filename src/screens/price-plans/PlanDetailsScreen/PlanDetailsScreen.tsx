import React, { useEffect } from 'react';
import { EmitterSubscription, View } from 'react-native';
import {
  ProductPurchase,
  PurchaseError,
  SubscriptionPurchase,
  finishTransaction,
  purchaseErrorListener,
  purchaseUpdatedListener,
  requestSubscription,
} from 'react-native-iap';

import { Button, PricePlanFeatureListTickIcon, Text } from '@sz/components';
import { tw } from '@sz/config';
import { Color, Route, TextAlignment, TextVariant } from '@sz/constants';
import { FinalPlanData } from '@sz/models';
import { NavigationService, ToastService } from '@sz/services';
import { useDispatch } from '@sz/stores';

import { BaseScreen } from './../../components';

const TEST_ID_PREFIX = 'PlanDetailsScreen';

export function PlanDetailsScreen({ route }) {
  const data = route.params.params.item as FinalPlanData;
  const dispatch = useDispatch();

  const onProceed = async (data: FinalPlanData) => {
    //Free user plan
    if (data.productId === '') {
      try {
        await dispatch.userStore.addSubscription({ planId: data.id });
        NavigationService.navigate(Route.HomeTab);
      } catch (error) {
        ToastService.error({
          message: 'Failed!',
          description: error.data.message,
        });
      }
      //paid user plans
    } else {
      try {
        await requestSubscription({ sku: data.productId });
      } catch (error) {
        ToastService.information({
          message: 'Alert!',
          description: 'Could not complete the payment!',
        });
      }
    }
  };

  useEffect(() => {
    const unsubscribePurchaseEvent: EmitterSubscription = purchaseUpdatedListener(
      async (purchase: ProductPurchase | SubscriptionPurchase) => {
        const receipt = purchase.transactionReceipt
          ? purchase.transactionReceipt
          : (purchase as unknown as { originalJson: string }).originalJson;

        if (receipt) {
          try {
            //TODO::need to validate this receipt before proceed
            await finishTransaction({ purchase });
            //TODO::persist the payment status.
            //TODO::handle worst case scenarios(app crashes, internet connection lost)
            NavigationService.navigate(Route.HomeTab);
          } catch (error) {
            ToastService.error({ message: 'Failed!', description: 'Something went wrong during the purchase!' }); //TODO::add a proper message
          }
        }
      },
    );

    return unsubscribePurchaseEvent?.remove();
  }, []);

  useEffect(() => {
    const unsubscribePurchaseFailEvent: EmitterSubscription = purchaseErrorListener((error: PurchaseError) => {
      ToastService.error({ message: 'Failed!', description: error.message });
    });

    return unsubscribePurchaseFailEvent?.remove();
  }, []);

  return (
    <BaseScreen testID={TEST_ID_PREFIX}>
      <View style={tw`flex-1 justify-between mx-4`}>
        <View style={tw`flex-1`}>
          <View style={tw`mt-2`}>
            <Text variant={TextVariant.Labels}>You have chosen the</Text>
          </View>
          <View style={tw`mt-1`}>
            <Text variant={TextVariant.SubTitle1}>{data.name}</Text>
          </View>
          <View style={tw`items-start`}>
            <View style={tw`mt-10`}>
              <Text variant={TextVariant.Body2Regular}>{data.description + ':'}</Text>
            </View>
            <View style={tw`mx-6 mt-3 items-start`}>
              {data.actions?.map(feature => {
                return (
                  <View style={tw`flex-row gap-2 mt-2`} key={`${TEST_ID_PREFIX}-Feature-${feature}`}>
                    <View style={tw`mt-1.5`}>
                      <PricePlanFeatureListTickIcon />
                    </View>
                    <Text textAlign={TextAlignment.Left} variant={TextVariant.Labels} color={Color.Neutral.Sz100}>
                      {feature}
                    </Text>
                  </View>
                );
              })}
            </View>
          </View>
        </View>
        <View style={tw`items-center mt-18 mb-5`}>
          <View style={tw`flex-row gap-2 items-center`}>
            <Text variant={TextVariant.Heading1} color={Color.Neutral.Sz100}>
              {`$${data.price}`}
            </Text>
            <View style={tw`mt-6`}>
              <Text variant={TextVariant.Body2Regular} color={Color.Neutral.Sz100}>
                {data.price === 0 ? '' : '/ ' + data.frequency}
              </Text>
            </View>
          </View>
          <View style={tw`mt-8 mb-3`}>
            <Button
              onPress={() => onProceed(data)}
              title={data.price === 0 ? 'next' : 'proceed to pay'}
              testID={`${TEST_ID_PREFIX}-PayButton`}
            />
          </View>
        </View>
      </View>
    </BaseScreen>
  );
}
