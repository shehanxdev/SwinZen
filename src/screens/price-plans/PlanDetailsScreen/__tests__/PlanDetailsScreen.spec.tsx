import { render } from '@testing-library/react-native';
import * as React from 'react';

import { PlanDetailsScreen } from '../PlanDetailsScreen';

describe('Plan Details Screen screen', () => {
  const testID = 'PlanDetailsScreen';
  const mockRoute = {
    params: {
      params: {
        item: {
          name: 'testName',
          description: 'testLongDescription',
          features: ['testFeatureOne', 'testFeatureTwo'],
          price: 99.0,
          frequency: 'testFrequency',
        },
      },
    },
  };
  const getRenderedScreen = () => render(<PlanDetailsScreen route={mockRoute} />);

  it(`should render Plan Details Screen correctly`, async () => {
    const renderer = getRenderedScreen();
    const renderTree = renderer.toJSON();
    expect(renderTree).toMatchSnapshot();
  });

  it('should find the Plan Details Screen via testID', () => {
    const { getByTestId } = getRenderedScreen();
    const foundScreen = getByTestId(testID);
    expect(foundScreen).toBeDefined();
  });
});
