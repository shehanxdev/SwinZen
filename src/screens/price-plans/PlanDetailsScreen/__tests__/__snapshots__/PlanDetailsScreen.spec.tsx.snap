// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plan Details Screen screen should render Plan Details Screen correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    {
      "flex": 1,
    }
  }
>
  <BVLinearGradient
    colors={
      [
        4279917603,
        4278984207,
        4278650631,
      ]
    }
    endPoint={
      {
        "x": 0.5,
        "y": 1,
      }
    }
    locations={null}
    startPoint={
      {
        "x": 0.5,
        "y": 0,
      }
    }
    style={
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
        "position": "relative",
      }
    }
    testID="PlanDetailsScreenTestID"
  >
    <Image
      resizeMode="repeat"
      source={
        {
          "testUri": "../../../src/assets/images/grassBackground.jpg",
        }
      }
      style={
        {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
          "opacity": 0.1,
          "position": "absolute",
        }
      }
    />
    <RCTSafeAreaView
      style={
        {
          "height": "100%",
          "paddingTop": 50,
        }
      }
    >
      <View
        onLayout={[Function]}
        style={
          [
            {
              "flexBasis": "0%",
              "flexGrow": 1,
              "flexShrink": 1,
            },
            {
              "paddingBottom": 0,
            },
          ]
        }
      >
        <RCTScrollView
          contentContainerStyle={
            {
              "flexGrow": 1,
            }
          }
          keyboardShouldPersistTaps="handled"
        >
          <View>
            <View
              style={
                {
                  "flexBasis": "0%",
                  "flexGrow": 1,
                  "flexShrink": 1,
                  "justifyContent": "space-between",
                  "marginLeft": 16,
                  "marginRight": 16,
                }
              }
            >
              <View
                style={
                  {
                    "flexBasis": "0%",
                    "flexGrow": 1,
                    "flexShrink": 1,
                    "marginLeft": 20,
                    "marginRight": 20,
                  }
                }
              >
                <View
                  style={
                    {
                      "marginTop": 8,
                    }
                  }
                >
                  <Text
                    style={
                      [
                        {
                          "color": "#F8F9FA",
                          "lineHeight": 24,
                        },
                        {
                          "fontFamily": "SourceSansPro-Regular",
                          "fontSize": 15,
                          "fontWeight": "600",
                          "textAlign": "center",
                        },
                      ]
                    }
                  >
                    You have chosen the
                  </Text>
                </View>
                <View
                  style={
                    {
                      "marginTop": 4,
                    }
                  }
                >
                  <Text
                    style={
                      [
                        {
                          "color": "#F8F9FA",
                          "lineHeight": 40,
                        },
                        {
                          "fontFamily": "SourceSansPro-Regular",
                          "fontSize": 28,
                          "fontWeight": "600",
                          "textAlign": "center",
                        },
                      ]
                    }
                  >
                    testName
                  </Text>
                </View>
                <View
                  style={
                    {
                      "marginTop": 40,
                    }
                  }
                >
                  <Text
                    style={
                      [
                        {
                          "color": "#F8F9FA",
                          "lineHeight": 24,
                        },
                        {
                          "fontFamily": "SourceSansPro-Regular",
                          "fontSize": 16,
                          "fontWeight": "normal",
                          "textAlign": "center",
                        },
                      ]
                    }
                  >
                    testLongDescription
                  </Text>
                </View>
                <View
                  style={
                    {
                      "marginTop": 16,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "gap": 8,
                        "justifyContent": "center",
                        "marginTop": 8,
                      }
                    }
                  >
                    <View
                      style={
                        {
                          "marginTop": 4,
                        }
                      }
                    >
                      <SvgMock
                        fill="#A2FD2F"
                        height={13}
                        width={13}
                      />
                    </View>
                    <Text
                      style={
                        [
                          {
                            "color": "#F8F9FA",
                            "lineHeight": 24,
                          },
                          {
                            "fontFamily": "SourceSansPro-Regular",
                            "fontSize": 15,
                            "fontWeight": "600",
                            "textAlign": "center",
                          },
                        ]
                      }
                    >
                      testFeatureOne
                    </Text>
                  </View>
                  <View
                    style={
                      {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "gap": 8,
                        "justifyContent": "center",
                        "marginTop": 8,
                      }
                    }
                  >
                    <View
                      style={
                        {
                          "marginTop": 4,
                        }
                      }
                    >
                      <SvgMock
                        fill="#A2FD2F"
                        height={13}
                        width={13}
                      />
                    </View>
                    <Text
                      style={
                        [
                          {
                            "color": "#F8F9FA",
                            "lineHeight": 24,
                          },
                          {
                            "fontFamily": "SourceSansPro-Regular",
                            "fontSize": 15,
                            "fontWeight": "600",
                            "textAlign": "center",
                          },
                        ]
                      }
                    >
                      testFeatureTwo
                    </Text>
                  </View>
                </View>
              </View>
              <View
                style={
                  {
                    "alignItems": "center",
                    "marginBottom": 20,
                  }
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "gap": 8,
                    }
                  }
                >
                  <Text
                    style={
                      [
                        {
                          "color": "#F8F9FA",
                          "lineHeight": 72,
                        },
                        {
                          "fontFamily": "SourceSansPro-Regular",
                          "fontSize": 60,
                          "fontWeight": "bold",
                          "textAlign": "center",
                        },
                      ]
                    }
                  >
                    $99
                  </Text>
                  <View
                    style={
                      {
                        "marginTop": 24,
                      }
                    }
                  >
                    <Text
                      style={
                        [
                          {
                            "color": "#F8F9FA",
                            "lineHeight": 24,
                          },
                          {
                            "fontFamily": "SourceSansPro-Regular",
                            "fontSize": 16,
                            "fontWeight": "normal",
                            "textAlign": "center",
                          },
                        ]
                      }
                    >
                      /testFrequency
                    </Text>
                  </View>
                </View>
                <View
                  style={
                    {
                      "marginBottom": 12,
                      "marginTop": 64,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "flexDirection": "row",
                      }
                    }
                  >
                    <View
                      collapsable={false}
                      onTouchEnd={[Function]}
                      onTouchStart={[Function]}
                      style={
                        {
                          "backgroundColor": "#A2FD2F",
                          "borderColor": "#ffffff00",
                          "borderRadius": 10,
                          "borderStyle": "solid",
                          "borderWidth": 1,
                          "elevation": 0,
                          "justifyContent": "center",
                          "minWidth": "auto",
                          "shadowColor": "#000",
                          "shadowOffset": {
                            "height": 0,
                            "width": 0,
                          },
                          "shadowOpacity": 0,
                          "shadowRadius": 0,
                          "width": "100%",
                        }
                      }
                    >
                      <View
                        accessibilityRole="button"
                        accessibilityState={
                          {
                            "disabled": false,
                          }
                        }
                        accessibilityValue={
                          {
                            "max": undefined,
                            "min": undefined,
                            "now": undefined,
                            "text": undefined,
                          }
                        }
                        accessible={true}
                        focusable={true}
                        onClick={[Function]}
                        onResponderGrant={[Function]}
                        onResponderMove={[Function]}
                        onResponderRelease={[Function]}
                        onResponderTerminate={[Function]}
                        onResponderTerminationRequest={[Function]}
                        onStartShouldSetResponder={[Function]}
                        style={
                          [
                            {
                              "overflow": "hidden",
                            },
                            {
                              "borderRadius": 10,
                            },
                          ]
                        }
                      >
                        <View
                          style={
                            [
                              {
                                "alignItems": "center",
                                "flexDirection": "row",
                                "justifyContent": "center",
                              },
                              {
                                "height": 48,
                              },
                            ]
                          }
                        >
                          <Text
                            numberOfLines={1}
                            selectable={false}
                            style={
                              [
                                {
                                  "color": "#000000",
                                  "fontFamily": "SourceSansPro-Regular",
                                  "fontWeight": "normal",
                                },
                                {
                                  "textAlign": "left",
                                },
                                [
                                  {
                                    "letterSpacing": 1,
                                    "marginHorizontal": 16,
                                    "marginVertical": 9,
                                    "textAlign": "center",
                                  },
                                  {
                                    "marginHorizontal": 8,
                                  },
                                  false,
                                  {
                                    "color": "#6200ee",
                                    "fontFamily": "SourceSansPro-SemiBold",
                                    "fontWeight": "normal",
                                  },
                                  {
                                    "fontFamily": "SourceSansPro-SemiBold",
                                    "fontWeight": "normal",
                                  },
                                  undefined,
                                ],
                              ]
                            }
                          >
                            <Text
                              style={
                                [
                                  {
                                    "color": "#283E36",
                                    "lineHeight": 28,
                                  },
                                  {
                                    "fontFamily": "SourceSansPro-SemiBold",
                                    "fontSize": 18,
                                    "fontWeight": "bold",
                                    "textAlign": "center",
                                  },
                                ]
                              }
                            >
                              PROCEED TO PAY
                            </Text>
                          </Text>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </RCTSafeAreaView>
  </BVLinearGradient>
</View>
`;
